### ISSUE TITLE
should catch mistakenly-doubled reserve words

### ISSUE THREAD

**Issue by radiovisual** [eslint/eslint#5983]
**What version of ESLint are you using?**
2.7.0

**What parser (default, Babel-ESLint, etc.) are you using?**
- parser: [XO](https://github.com/sindresorhus/xo)

**Please show your full configuration:**
- configuration: [eslint-config-xo](https://github.com/sindresorhus/eslint-config-xo/blob/master/index.js)

**What did you do? Please include the actual source code causing the issue.**

Consider this block of code, which mistakenly doubles-up the keywords `import` and `new`:
```js
import import OAuthError from 'node-oauth-error';
const date = new new Date();
console.log(date);
```
ESLint correctly marks the double `import import` as an error, but ignores the doubled `new new` tokens with the following output:

```
1:8  error  Parsing error: Unexpected token
```

**What did you expect to happen?**

I would expect ESLint to see `new new` in the same way it sees `import import`

**Comment by platinumazure** [comment]
Unfortunately, ESLint can't flag issues that are caught by the parser, because ESLint operates on an abstract syntax tree (which is the output of the parser). So if the parser fails, there is no AST, and ESLint cannot run on a file.

I don't think there's much we can do here. @nzakas?

**Comment by radiovisual** [comment]
@platinumazure , thanks for the feedback. Though, I don't think the parsing fails, since the linting process runs perfectly and exits normally. The problem I am raising here is that even though the parsing isn't failing, the linting process is ignoring `new new` while seeing `import import` as a violation.

**Comment by nzakas** [comment]
@radiovisual the parser is failing at `import import` and so does not continue parsing after that point. If you fix that, it will then stop parsing at `new new`.

You can when ESLint is reporting a parser error rather than a linting error because the message is prefixed with "Parser Error".

As @platinumazure mentioned, there's unfortunately not anything we can do about this.

**Comment by radiovisual** [comment]
@nzakas , @platinumazure , This issue should be reopened (sorry for the confusion). Now I understand what you meant about the failed parsing at `import import`, however, I only had the `import import` next to the `new new` in my previous code sample for comparison to highlight the problem (where `import import` was being caught, but the `new new` was being ignored). The problem, however, is not in the failed parsing, because if you run ESLint on the code below, then ESLint passes without any errors at all, even though there are two `new new` keywords:

```js
const date = new new Date();
console.log(date);
```

So while ESLint sees `import import` as a problem, it doesn't see `new new` as a problem.

**Comment by radiovisual** [comment]
Below is a screenshot of my test run at http://eslint.org/demo/. Notice how the code sample passes as `Lint-free!` even though there is a `new new` in there:

The screenshot also show the changes I made to the rules and environments. The only changes I made to the defaults are:

- [x] node
- [x] es6
- [ ] no-console

![new-new](https://cloud.githubusercontent.com/assets/5614571/14879309/4e6f0ad2-0d20-11e6-9f06-48957525a84f.png)

**Comment by mysticatea** [comment]
@radiovisual `new new foo` is valid syntax in JavaScript.

![image](https://cloud.githubusercontent.com/assets/1937871/14879526/b6ab3ac8-0d64-11e6-8705-ef7502019d4a.png)

https://tc39.github.io/ecma262/#prod-NewExpression

**Comment by radiovisual** [comment]
@mysticatea , ah yes! You are correct sir! That changes everything. Thanks for the clarification!

**Comment by platinumazure** [comment]
I could see a use case for a rule that flags a "double new" since I assume
that would be unintentional the vast majority of the time.

**Comment by radiovisual** [comment]
@platinumazure , I agree. I was actually going to recommend the same thing. I have been writing Javascript for a while now, but I have never seen/used the double new syntax, so from my perspective, if a double-new were to creep its way into my code, I would most likely not be using it according to it's intended syntax, and could therefore manifest some erratic behaviors in my application, or confuse the intention of the code to anyone reading it.
